nexus/src/app/quiesce.rs

```diff
@@ -201,10 +294,222 @@ async fn do_quiesce(quiesce: NexusQuiesceHandle) {
+        );
+        error!(
+            log,
+            "try_saga_quiesce(): impossible condition";
+            InlineErrorChain::new(&*error)
+        );
+        return Err(error);
+    };
+
+    // Wait a bounded amount of time for sagas to become drained as of this
+    // blueprint.
+    let Ok(_) = tokio::time::timeout(
+        pause_timeout,
+        saga_quiesce.wait_for_drained_blueprint(current_blueprint.id),
+    )
+    .await
+    else {
+        // We're still not drained as of this blueprint (or we're drained as of
+        // a newer one).  It's possible that the current target blueprint could
+        // have changed, so bail out and let the caller invoke us again.
+        bail!("not locally drained as of blueprint {}", current_blueprint.id);
+    };
+
+    // We're drained up through this blueprint.  First, update our record to
+    // reflect that, if we haven't already.
+    match last_recorded_blueprint_id {
+        Some(blueprint_id) if *blueprint_id == current_blueprint.id => (),
+        _ => {
+            info!(
+                log,
+                "locally drained as of blueprint";
+                "blueprint_id" => %current_blueprint.id
+            );
+            let try_update = datastore
+                .database_nexus_access_update_blueprint(
+                    quiesce_opctx,
+                    my_nexus_id,
+                    Some(current_blueprint.id),
+                )
+                .await;
+            match try_update {
+                Err(error) => {
+                    return Err(anyhow!(error).context(
+                        "updating our db_metadata_nexus record's drained \
+                         blueprint",
+                    ));
+                }
+                Ok(0) => (),
+                Ok(count) => {
+                    info!(
+                        log,
+                        "updated our db_metadata_nexus record blueprint id";
+                        "nexus_id" => %my_nexus_id,
+                        "blueprint_id" => %current_blueprint.id,
+                        "count" => count,
+                    );
+                }
+            }
+
+            *last_recorded_blueprint_id = Some(current_blueprint.id);
+        }
+    };
+
+    // Now, see if everybody else is drained up to the same point, or if any
+    // of them is already quiesced.  That would mean *they* already saw that we
+    // were all drained up to the same point, which is good enough for us to
+    // proceed, too.
+    let other_nexus_ids: BTreeSet<OmicronZoneUuid> = current_blueprint
+        .all_omicron_zones(BlueprintZoneDisposition::is_in_service)
+        .filter_map(|(_sled_id, zone)| {
+            if let BlueprintZoneType::Nexus(nexus) = &zone.zone_type {
+                (nexus.nexus_generation == my_generation).then_some(zone.id)
+            } else {
+                None
+            }
+        })
+        .collect();
+    assert!(!other_nexus_ids.is_empty());
```

Comment:
At first glance I thought this assert was wrong because we might have no other Nexuses, but IIUC `other_nexus_ids` also contains _our_ ID, right? Could we rename this? Not sure what to suggest - maybe `our_gen_nexus_ids`?